src/pudl/output/ferc1.py

```diff
@@ -1731,68 +1842,135 @@ def exploded_calcs_to_digraph(
+        some error checking to try and ensure that the weights and tags that are being
+        associated with the forest are internally self-consistent.
+
+        We check whether there are multiple different weights assocated with the same
+        node in the calculation components. There are a few instances where this is
+        expected, but if there a lot of conflicting weights something is probably wrong.
+
+        We check whether any of the nodes that were orphaned (never connected to the
+        graph) or that were pruned in the course of enforcing a forest structure had
+        manually assigned tags (e.g. indicating whether they contribute to rate base).
+        If they do, then the final exploded data table may not capture all of the
+        manually assigned metadata, and we either need to edit the metadata, or figure
+        out why those nodes aren't being included in the final calculation forest.
+        """
+        annotated_forest = deepcopy(self.forest)
+        nx.set_node_attributes(annotated_forest, self.node_attrs)
+        nx.set_edge_attributes(annotated_forest, self.edge_attrs)
+
+        logger.info("Checking whether any pruned nodes were also tagged.")
+        self.check_lost_tags(lost_nodes=self.pruned)
+        logger.info("Checking whether any orphaned nodes were also tagged.")
+        self.check_lost_tags(lost_nodes=self.orphans)
+        self.check_conflicting_tags(annotated_forest)
+        return annotated_forest
+
+    def check_lost_tags(self: Self, lost_nodes: list[NodeId]) -> None:
+        """Check whether any of the input lost nodes were also tagged nodes."""
+        if lost_nodes:
+            lost = pd.DataFrame(lost_nodes).set_index(self.calc_cols)
+            tagged = self.tags.set_index(self.calc_cols)
+            lost_tagged = tagged.index.intersection(lost.index)
+            if not lost_tagged.empty:
+                logger.warning(
+                    "The following tagged nodes were lost in building the forest:\n"
+                    f"{tagged.loc[lost_tagged].sort_index()}"
+                )
 
-    @property
+    @staticmethod
+    def check_conflicting_tags(annotated_forest: nx.DiGraph) -> None:
+        """Check for conflicts between ancestor and descendant tags.
+
+        At this point, we have just applied the manually compiled tags to the nodes in
+        the forest, and haven't yet propagated them down to the leaves. It's possible
+        that ancestor nodes (closer to the roots) might have tags associated with them
+        that are in conflict with descendant nodes (closer to the leaves). If that's
+        the case then when we propagate the tags to the leaves, whichever tag is
+        propagated last will end up taking precedence.
+
+        These kinds of conflicts are probably due to errors in the tagging metadata, and
+        should be investigated.
+        """
+        nodes = annotated_forest.nodes
+        for ancestor in nodes:
```

Comment:
This is some deep nesting, ha. I don't really see an obvious way to deal with it beyond moving the nesting elsewhere... So this seems fine. I guess you could define a few little helper functions inline here.

Reply:
It's not great... but it checks the thing and it's isolated off here by itself.